### DISCUSSION TITLE
add support for OIDC picture claim

### DISCUSSION THREAD

**Discussion by UiP9AV6Y** [authelia/authelia#4812] [Ideas]
### Description

the OpenID profile scope contains a picture claim. support for this was already requested in #4295, which was closed in favour of #2868. the latter however is a feature request for mapping arbitrary backend fields to OIDC claims. the *jpegPhoto* (or *thumbnailPhoto* in AD) is stored as binary data, while the OIDC [claim](https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.5.1) requires an URL pointing to a photo; those two are not mappabale 1:1

if serving pictures is not in the scope of authelia, support for [libravatar](https://wiki.libravatar.org/api/) (an open alternative with fallback to gravatar) would probably also suffice; authelia already has access to the mail attribute. in that case i would open a new issue to discuss the implementation of this feature there.

### Use Case

displaying a user avatar image in applications using authelia as identity provider

### Details

in order to support this, authelia would need to serve the image data from the auth backend itself and announce it via the *photo* claim. the URL format can be arbitrary, as long as it is unique. the uid would probably cover this requirement, as it is already used for authentication (e.g. `https://auth.example.com/api/oidc/pictures/$URL_ENCODED_UID`).

the file-based auth backend would also store image data in its database as base64-encoded raw image data.

### Documentation

OpenID Standard Claims: https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.5.1

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)
- [X] I have checked for related issues and checked the documentation

**Comment by james-d-elliott** [comment]
> the latter however is a feature request for mapping arbitrary backend fields to OIDC claims. the jpegPhoto (or thumbnailPhoto in AD) is stored as binary data, while the OIDC [claim](https://openid.net/specs/openid-connect-core-1_0.html#rfc.section.5.1) requires an URL pointing to a photo; those two are not mappabale 1:1

Yes. This was acknowledged in the issue and is completely irrelevant nor was it why it is a duplicate of #2868. In fact #2868 allows and encourages admins to configure this in a spec conformant way as the URL can easily be stored in an attribute for the users the admin has consciously decided to configure. This is why it was closed.

> in order to support this, authelia would need to serve the image data from the auth backend itself and announce it via the photo claim.

This is not a priority for us at the moment and there are significant features that need to predate this such as an admin UI for an admin to upload said images.

> the URL format can be arbitrary, as long as it is unique. the uid would probably cover this requirement, as it is already used for authentication (e.g. https://auth.example.com/api/oidc/pictures/$URL_ENCODED_UID).

The format if supported will be `/path/<sub>.jpeg` where `<sub>` is the `sub` claim.

**Reply by UiP9AV6Y** [reply]
> ... as the URL can easily be stored in an attribute for the users the admin has consciously decided to configure

yes, i could store an arbitrary attribute in LDAP containing an URL and then use the mapping feature (when authelia has implemented it) to expose it as *picture claim*. however, i am specifically talking about the [jpegPhoto](https://www.rfc-editor.org/rfc/rfc2798#section-2.6) attribute of the *inetOrgPerson* LDAP Object class, which is specified as being a binary attribute (as it is image data stored directly in the LDAP tree). this attribute can **not** be mapped to the *picture* claim, as the OIDC specification dictates this attribute to be an URL.

> This is not a priority for us at the moment and there are significant features that need to predate this such as an admin UI for an admin to upload said images.

i was not expecting Authelia to have authority over this attribute, but simply act as proxy to serve the binary data directly from LDAP (optionally with caching) as `image/jpeg` (similar to how [canaille does it](https://gitlab.com/yaal/canaille/-/blob/4f23dd07dfa2696663960ce628f7e302a9460280/canaille/account.py#L626)). the image data itself is managed the same way all other data in LDAP is managed; either by an administrator or through a self-service, but in any case something outside of the scope of Authelia
